File: golife/board.go
package main

import (
	"math"
	"sync"
)

type Board struct {
	// gophers    int
	sideLength int
	Boards     [2][]int
}

func isPower2(n int) bool {
	if n < 1 {
		return false
	}

	for n > 1 {
		if n%2 == 1 {
			return false
		}
		n /= 2
	}
	return true
}

func CreateBoard(sideLength int) Board {
	//check if gophers and sideLength are powers of two
	/*
		if !isPower2(gophers) {
			panic("ERROR: Number of workers MUST be a power of 2!")
		}

		if gophers > sideLength*sideLength {
			panic("ERROR: More Goroutines than BOARD CELLS")
		}
	*/
	boards := [...][]int{make([]int, sideLength*sideLength), make([]int, sideLength*sideLength)}
	return Board{
		// gophers:    gophers,
		sideLength: sideLength,
		Boards:     boards,
	}
}

// linha i coluna j
func (b *Board) numNeighbors(i, j int) int {
	n := 0
	size := b.sideLength
	for y := -1; y <= 1; y++ {
		yi := (i + y + b.sideLength) % b.sideLength
		for x := -1; x <= 1; x++ {
			if x == 0 && y == 0 {
				continue
			}

			xj := (x + j + b.sideLength) % b.sideLength

			n += b.Boards[0][yi*size+xj]
		}
	}

	return n
}

// returns number of alive cells in table!
// linha i coluna j
func (b *Board) subTableWork(i, j, height, width int) int {
	alive := 0
	side := b.sideLength

	for y := i; y < i+height; y++ {
		for x := j; x < j+width; x++ {
			nn := b.numNeighbors(y, x)

			if b.Boards[0][y*side+x] == 1 {
				if nn < 2 || nn > 3 {
					//se vivo, mas menos de 2 ou mais de 3 vizinhos, morre
					b.Boards[1][y*side+x] = 0
				} else {
					b.Boards[1][y*side+x] = 1
					alive++
				}

			} else {
				if nn == 3 {
					//se morto, mas com 3 vizinhos, vive
					b.Boards[1][y*side+x] = 1
					alive++
				} else {
					b.Boards[1][y*side+x] = 0
				}
			}
		}
	}
	return alive
}

func (b *Board) swapBoards() {
	b.Boards[0], b.Boards[1] = b.Boards[1], b.Boards[0]
}

// inicialmente, maneira mais boba possivel, divide o tabuleiro entre os gophers
func (b *Board) IterateSequentially() {
	b.subTableWork(0, 0, b.sideLength, b.sideLength)
	b.swapBoards()
}

// Argument MUST be a power of 2!
func (b *Board) determineGopherTableSize(gophers int) (int, int) {
	if !isPower2(b.sideLength) || !isPower2(gophers) {
		return 1, b.sideLength
	}

	sqrRoot := int(math.Sqrt(float64(gophers)))
	width, height := b.sideLength, b.sideLength
	if sqrRoot*sqrRoot != int(gophers) { // i.e gophers is of the form 2 ^ (2k + 1)
		height /= 2
		sqrRoot = int(math.Sqrt(float64(gophers / 2)))
	}

	width /= sqrRoot
	height /= sqrRoot

	if width == 0 {
		width = 1
	}
	if height == 0 {
		height = 1
	}

	return height, width
}

// The number of goroutines must be a power of 2
func (b *Board) IterateConcurrently(gophers int) {
	// test if number of gophers is square!
	width, height := b.determineGopherTableSize(gophers)

	var wg sync.WaitGroup
	for y := 0; y < b.sideLength; y += height {
		for x := 0; x < b.sideLength; x += width {
			wg.Add(1)

			//asks for a goroutine to do a table!
			go func(i, j int) {
				b.subTableWork(i, j, height, width)
				wg.Done()
			}(y, x)

		}
	}

	//wait for all goroutines!
	wg.Wait()
	b.swapBoards()
}

func (b *Board) IterateConcurrentlyByLine(gophers int) {
	// test if number of gophers is square!
	var wg sync.WaitGroup
	for y := 0; y < b.sideLength; y++ {
		wg.Add(1)
		go func(i, j int) {
			b.subTableWork(i, j, 1, b.sideLength)
			wg.Done()
		}(y, 0)
	}

	//wait for all goroutines!
	wg.Wait()
	b.swapBoards()
}

File: c-life/src/board.c
#include <stdlib.h>
#include <stdio.h>
#include "board.h"

board_t *create_board(int sideLength){
    board_t *b = malloc(sizeof(board_t));
    b->boards[0] = malloc(sizeof(int)*sideLength*sideLength);
    b->boards[1] = malloc(sizeof(int)*sideLength*sideLength);
    b->sideLenght = sideLength;
    return b;
}

int num_neighbors(board_t* b, int x, int y){
    int num_neighbors = 0;
    for(int i = -1; i <= 1; i++){
        for(int j = -1; j <= 1; j++){
            if(i == 0 && j == 0) continue;
            int yi = (y + i + b->sideLenght) % b->sideLenght;
            int xj = (x + j + b->sideLenght) % b->sideLenght;
            num_neighbors += (b->boards[0])[yi * b->sideLenght + xj];
        }
    }
    return num_neighbors;
}

int subtable_work(board_t *b, int x, int y, int width, int height){
    int alive = 0;
    int side = b->sideLenght; // just to diminish size
    for(int i = y; i < y + height; i++){
        for(int j = x; j < x + width; j++){
            int nn = num_neighbors(b,j,i);
            if((b->boards[0])[i*side + j]){
                if(nn < 2 || nn > 3) {
                    (b->boards[1])[i*side + j] = 0;
                } else {
                    (b->boards[1])[i*side + j] = 1;
                    alive++;
                }
            } else {
                if (nn == 3){
                    b->boards[1][i*side + j] = 1;
                    alive++;
                } 
                else {
                    b->boards[1][i*side + j] = 0;
                }
            }
        }
    }
    return alive;
}

void swap_boards(board_t *b){
    int* cup = b->boards[0];
    b->boards[0] = b->boards[1];
    b->boards[1] = cup;
}

void iterate_sequentially(board_t *b){
    subtable_work(b,0,0,b->sideLenght,b->sideLenght);
    swap_boards(b);
}

void destroy_board(board_t *b){
    free(b->boards[0]);
    free(b->boards[1]);
}